### PR TITLE
[2267] Support auto ELK layout with React Flow.

### LINKED ISSUE #2267: Support automatically diagrams with an automatic layout
The ELK algorithm should be called after each new version of the diagram received.

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/2267

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/representation/DiagramRepresentation.types.ts [hunk 1]
```diff
@@ -12,11 +12,12 @@
  *******************************************************************************/
 
 import { GQLDiagramEventPayload } from '../graphql/subscription/diagramEventSubscription.types';
-import { Diagram } from '../renderer/DiagramRenderer.types';
+import { Diagram, DiagramDescription } from '../renderer/DiagramRenderer.types';
 
 export interface DiagramRepresentationState {
   id: string;
   diagram: Diagram | null;
+  diagramDescription: DiagramDescription | null;
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/representation/DiagramRepresentation.types.ts]: Do we really want `diagramDescription` to be `null`?

Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/representation/DiagramRepresentation.types.ts]: It seems you are just using `autoLayout`, why do you need the whole diagram description here?
Maybe `autoLayout` could be enough.

Comment (gcoutable) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/representation/DiagramRepresentation.types.ts]: yes, it can be `null` here

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.types.ts [hunk 2]
```diff
@@ -17,10 +17,17 @@ import { CustomEdgeData } from './edge/CustomEdge.types';
 
 export interface DiagramRendererProps {
   diagram: Diagram;
+  diagramDescription: DiagramDescription | null;
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.types.ts]: Do we really want `diagramDescription` to be `null`?

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx [hunk 3]
```diff
@@ -57,7 +57,7 @@ const edgeTypes: EdgeTypes = {
 
 const isSelectChange = (change: NodeChange): change is NodeSelectionChange => change.type === 'select';
 
-export const DiagramRenderer = ({ diagram, selection, setSelection }: DiagramRendererProps) => {
+export const DiagramRenderer = ({ diagram, autoLayoutElk, selection, setSelection }: DiagramRendererProps) => {
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx]: just `autoLayout`

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.types.ts [hunk 4]
```diff
@@ -17,10 +17,10 @@ import { CustomEdgeData } from './edge/CustomEdge.types';
 
 export interface DiagramRendererProps {
   diagram: Diagram;
+  autoLayoutElk: boolean;
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.types.ts]: just `autoLayout`

**Review by gcoutable — CHANGES_REQUESTED** [review]

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.types.ts (hunk 4 above)
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.types.ts]: There is already a function with this name 
` const { autoLayout } = useLayout();`

Comment (gcoutable) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.types.ts]: `isAutoLayout` then :)

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/representation/DiagramRepresentation.tsx [hunk 5]
```diff
@@ -59,32 +71,58 @@ export const DiagramRepresentation = ({
...
-      }
-    }
-  };
+  const {
+    loading: diagramDescriptionLoading,
+    data: diagramDescriptionData,
+    error: diagramDescriptionError,
+    subscribeToMore,
+  } = useQuery<GQLGetDiagramDescriptionData>(getDiagramDescriptionQuery, {
+    fetchPolicy: 'network-only',
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/representation/DiagramRepresentation.tsx:82]: Why did you change to 'network-only', 'no-cache' does not work anymore?

Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/representation/DiagramRepresentation.tsx:82]: no-cache does not work with subscribeToMore.

Comment (gcoutable) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/representation/DiagramRepresentation.tsx:82]: There might be an issue then. I have a warning after updating the diagram, sometimes an error.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/useArrangeAll.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+export const useArrangeAll = (): UseArrangeAllValue => {
+  const { autoLayout } = useLayout();
+  const { setNodes } = useReactFlow();
+
+  const layoutDiagram = (nodes: Node[], edges: Edge[]) => {
+    autoLayout(nodes, edges).then(({ nodes }) => {
+      setNodes(nodes);
+    });
+  };
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/layout/useArrangeAll.ts]: Put this in a `useCallback` with empty dependencies

**Review by gcoutable — CHANGES_REQUESTED** [review]

**Comment by gcoutable** [comment]
There are still two issues to fix:
- Sometimes an error replaces the diagram: https://github.com/eclipse-sirius/sirius-web/issues/2463
- The fit-to-screen is not applied after the auto-laid-out diagram is opened: https://github.com/eclipse-sirius/sirius-web/issues/2464
